### PR TITLE
Fuel type list

### PR DESCRIPTION
**Pull request by andrea-williams**
Not the only PR that will result from ticket #1555 - this PR only addresses the backend changes necessary to create a new endpoint so that all fuel types stored in our database will be retrieved when the HFI Calc page loads.
(You should be able to see this network call & response when you load HFI Calc)
# Test Links:
[Percentile Calculator](https://wps-pr-1892.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-1892.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-1892.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-1892.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-1892.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=1108&f=c5&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN#state=2ec784ca-c46a-49d0-b2b3-1cf32a9015a2&session_state=7d9447c8-db66-4661-b4cb-03d2ac0d1d8f&code=32292df4-2bdf-4f90-a4a8-c8dbcda682a9.7d9447c8-db66-4661-b4cb-03d2ac0d1d8f.2b63f390-f3dc-43ae-89f2-016453863476)
[Fire Behaviour Advisory](https://wps-pr-1892.apps.silver.devops.gov.bc.ca/fire-behaviour-advisory)
[HFI Calculator](https://wps-pr-1892.apps.silver.devops.gov.bc.ca/hfi-calculator)
[FWI Calculator](https://wps-pr-1892.apps.silver.devops.gov.bc.ca/fwi-calculator)

### CODE REVIEW

File: api/app/routers/hfi_calc.py [hunk 1]
```diff
@@ -140,6 +141,36 @@ def save_request_in_database(request: HFIResultRequest, username: str) -> bool:
     return False
 
 
+def extract_selected_stations(request: HFIResultRequest) -> List[int]:
```
Comment (Sybrand) [inline api/app/routers/hfi_calc.py:144]: Remove this function please (extract_selected_stations)

File: api/app/db/crud/hfi_calc.py [hunk 2]
```diff
@@ -78,3 +78,8 @@ def get_fire_centre_fire_start_ranges(session: Session, fire_centre_id: id) -> C
 def get_fire_start_lookup(session: Session) -> CursorResult:
     """ Get the fire start lookup table """
     return session.query(FireStartLookup)
+
+
+def get_fuel_types_from_db(session: Session) -> CursorResult:
+    """ Get the fuel types table  """
+    return session.query(FuelType)
```
Comment (Sybrand) [inline api/app/db/crud/hfi_calc.py:85]: We need to sort the fuel types. Hopefully we can get away with alphabetic for now.

**Review by Sybrand — CHANGES_REQUESTED** [review]

File: web/src/api/hfiCalculatorAPI.ts [hunk 3]
```diff
@@ -7,6 +7,47 @@ import {
 import { DateTime } from 'luxon'
 import { formatISODateInPST } from 'utils/date'
 
+export interface FuelType {
+  abbrev: string
```
Comment (conbrad) [inline web/src/api/hfiCalculatorAPI.ts:11]: Needs an `id` field too, yeah?

File: web/src/api/hfiCalculatorAPI.ts [hunk 4]
```diff
@@ -59,6 +107,12 @@ export async function loadDefaultHFIResult(
   return { ...data, planning_area_hfi_results: buildResult(data) }
 }
 
+export async function getFuelTypes(): Promise<FuelTypesResponse> {
+  const data = await axios.get<FuelTypesResponse>(baseUrl + 'fuel_types')
+  console.log(data)
```
Comment (conbrad) [inline web/src/api/hfiCalculatorAPI.ts:112]: console.log

**Review by conbrad — CHANGES_REQUESTED** [review]
Thanks for cleaning up the code. Just some minor comments.

File: web/src/api/hfiCalculatorAPI.ts (hunk 3 above)
Comment (andrea-williams) [reply web/src/api/hfiCalculatorAPI.ts:12]: It will do in the next PR - I've added it in now anyway

**Review by conbrad — APPROVED** [review]

File: web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts [hunk 5]
```diff
@@ -137,6 +145,10 @@ const dailiesSlice = createSlice({
       state.loading = false
       state.changeSaved = false
     },
+    fetchFuelTypesFailed(state: HFICalculatorState, action: PayloadAction<string>) {
+      state.error = action.payload
```
Comment (Sybrand) [inline web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts:149]: What happens when station load fails, but fuel type load works. Does the error message stick around? Does the loading indicator stop when one of those calls comes back? (Should be in loading state until both responses have come back).

File: web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts [hunk 6]
```diff
@@ -156,17 +168,27 @@ const dailiesSlice = createSlice({
       state.result = action.payload
       state.dateRange = action.payload?.date_range
       state.loading = false
+    },
+    setFuelTypes: (
+      state: HFICalculatorState,
+      action: PayloadAction<FuelTypesResponse>
+    ) => {
+      state.fuelTypes = action.payload.fuel_types
+      state.loading = false
```
Comment (Sybrand) [inline web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts:177]: Fuel types would then set load to false even though stations could still be loading.

Comment (Sybrand) [reply web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts:177]: Also this should be covered by tests.

File: web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts [hunk 7]
```diff
@@ -227,6 +249,17 @@ export const fetchGetPrepDateRange =
     }
   }
 
+export const fetchFuelTypes = (): AppThunk => async dispatch => {
+  try {
+    dispatch(fetchFuelTypesStart())
+    const fuelTypes = await getFuelTypes()
+    dispatch(setFuelTypes(fuelTypes))
```
Comment (Sybrand) [inline web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts:256]: Why is this not being covered by tests?

**Review by Sybrand — CHANGES_REQUESTED** [review]

File: web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts (hunk 7 above)
Comment (andrea-williams) [reply web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts:256]: I was leaning heavily into Agile Principle #10?
(Actually, I was just too focussed on updating the existing unit tests and forgot to write any new ones for this new call)

File: api/app/db/crud/hfi_calc.py [hunk 8]
```diff
@@ -78,3 +78,8 @@ def get_fire_centre_fire_start_ranges(session: Session, fire_centre_id: id) -> C
 def get_fire_start_lookup(session: Session) -> CursorResult:
     """ Get the fire start lookup table """
     return session.query(FireStartLookup)
+
+
+def get_fuel_types_from_db(session: Session) -> CursorResult:
```
Comment (Sybrand) [inline api/app/db/crud/hfi_calc.py:83]:
```suggestion
def get_fuel_types(session: Session) -> CursorResult:
```

I think "from_db" is a bit redundant - it's in app.db.crud.hfi_calc - so I think in that context it's clear.

Comment (andrea-williams) [reply api/app/db/crud/hfi_calc.py:95]: I agree "from_db" is redundant, but it was the quickest solution I could think of. In routers.hfi_calc, the /fuel_types endpoint's method is called "get_fuel_types", and when I'd previously called the crud method "get_fuel_types" too, there was an infinitely recursive loop in routers.hfi_calc where that file's "get_fuel_types" was calling itself.

Do you know how to get around that?

Comment (Sybrand) [reply api/app/db/crud/hfi_calc.py:95]: For sure. Either do an "from blah import thing as moo" or do an "import blah", and the call it by saying blah.thing.

Comment (andrea-williams) [reply api/app/db/crud/hfi_calc.py:95]: If you're wondering - yes, I am embarassed that I asked that. I don't know why I didn't think of that.

File: api/app/routers/hfi_calc.py [hunk 9]
```diff
@@ -140,6 +141,25 @@ def save_request_in_database(request: HFIResultRequest, username: str) -> bool:
     return False
 
 
+@router.get("/fuel_types")
+async def get_fuel_types(response: Response) -> FuelTypesResponse:
```
[the remark this answers is not in the document]
Comment (Sybrand) [reply api/app/routers/hfi_calc.py:145]: There still isn't coverage for this?

File: web/src/features/hfiCalculator/pages/HfiCalculatorPage.tsx [hunk 10]
```diff
@@ -72,15 +73,17 @@ const HfiCalculatorPage: React.FunctionComponent = () => {
   const dispatch: AppDispatch = useDispatch()
   const {
     fireCentres,
-    error: fireCentresError,
-    loading: fireCentresLoading
+    error: fireCentresError
+    // loading: fireCentresLoading
```
Comment (Sybrand) [inline web/src/features/hfiCalculator/pages/HfiCalculatorPage.tsx:77]: Do we need this comment?

**Review by conbrad — CHANGES_REQUESTED** [review]

File: web/src/features/hfiCalculator/slices/hfiCalculatorSlice.test.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  initialState,
+  fetchFuelTypesStart,
+  fetchFuelTypesFailed,
+  pdfDownloadEnd,
+  pdfDownloadStart,
+  getHFIResultFailed
+} from 'features/hfiCalculator/slices/hfiCalculatorSlice'
+
+describe('hfiCalculatorSlice', () => {
+  describe('reducer', () => {
```
[the remark this answers is not in the document]
Comment (andrea-williams) [reply web/src/features/hfiCalculator/slices/hfiCalculatorSlice.test.ts:11]: @conbrad could you clarify pls? In my view it has been addressed - I've removed the `loading` bool to use separate `pdfLoading` and `fuelTypesLoading` flags instead, which are being tested in this file

File: api/app/tests/hfi/test_hfi.py [hunk 12]
```diff
@@ -229,6 +236,31 @@ def test_valid_daily():
...
+    monkeypatch.setattr(app.routers.hfi_calc, 'crud_get_fuel_types', mock_get_fuel_types)
+    correct_response_file = os.path.join(os.path.dirname(__file__), 'test_valid_fuel_types_response.json')
+
+    client = TestClient(starlette_app)
+    response = client.get('/api/hfi-calc/fuel_types')
+    assert response.status_code == 200
+    assert response.headers['content-type'] == 'application/json'
+    with open(correct_response_file) as file_reader:
+        correct_response = file_reader.read()
+        correct_response = correct_response.replace('\n', '').replace(' ', '')
```
Comment (Sybrand) [inline api/app/tests/hfi/test_hfi.py:260]: It's better to not do a text comparison, then you don't have to do these adjustments. Rather load from file using `json.load` and then you can compare it to `response['response'].json()`

**Review by Sybrand — CHANGES_REQUESTED** [review]

File: web/cypress/integration/hfi-calculator-page.spec.ts [hunk 13]
```diff
@@ -78,8 +78,12 @@ describe('HFI Calculator Page', () => {
       cy.intercept('GET', 'api/hfi-calc/fire-centres', {
         fixture: 'hfi-calc/fire_centres.json'
       }).as('getFireCentres')
+      cy.intercept('GET', 'api/hfi-calc/fuel_types', {
```
Comment (Sybrand) [inline web/cypress/integration/hfi-calculator-page.spec.ts:81]: I think we should throw this inside interceptLoad - we can pass interceptLoad two json paths. Sonarcloud is legitimately picking this up as duplicate code.

**Review by conbrad — APPROVED** [review]
Once Sy's happy on this one, I'm happy

**Comment by Sybrand** [comment]
> Once Sy's happy on this one, I'm happy

Hey! That's my line!

File: api/app/tests/hfi/test_hfi.py [hunk 14]
```diff
@@ -229,6 +237,29 @@ def test_valid_daily():
...
+
+    monkeypatch.setattr(app.routers.hfi_calc, 'crud_get_fuel_types', mock_get_fuel_types)
+    correct_response_file = os.path.join(os.path.dirname(__file__), 'test_valid_fuel_types_response.json')
+
+    client = TestClient(starlette_app)
+    response = client.get('/api/hfi-calc/fuel_types')
+    assert response.status_code == 200
+    assert response.headers['content-type'] == 'application/json'
+    with open(correct_response_file) as file_reader:
+        assert json.loads(response.text) == json.loads(file_reader.read())
```
Comment (Sybrand) [inline api/app/tests/hfi/test_hfi.py:260]:
```suggestion
        assert response.json() == json.load(file_reader.read())
```

**Review by Sybrand — APPROVED** [review]
Minor suggested change on json assertion, otherwise LGTM.

**Pull request merged**